pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -234,3 +234,53 @@ func GetPipelineConditionStatus(prName string, state []*ResolvedPipelineRunTask,
 		Message: "All Tasks have completed executing",
 	}
 }
+
+func findReferencedTask(pb string, state []ResolvedPipelineRunTask) *ResolvedPipelineRunTask {
+	var depTask *ResolvedPipelineRunTask
+	for _, rprtRef := range state {
+		if rprtRef.PipelineTask.Name == pb {
+			depTask = &rprtRef
+			break
+		}
+	}
+	return depTask
+}
+
+// ValidateProvidedBy will look at any `providedBy` clauses in the resolved PipelineRun state
+// and validate it: the `providedBy` must specify an input of the current `Task`. The `PipelineTask`
+// it is provided by must actually exist in the `Pipeline`. The `PipelineResource` that is bound to the input
+// must be the same `PipelineResource` that was bound to the output of the previous `Task`. If the state is
+// not valid, it will return an error.
+func ValidateProvidedBy(state []ResolvedPipelineRunTask) error {
+	for _, rprt := range state {
+		for _, dep := range rprt.PipelineTask.ResourceDependencies {
+			inputBinding, ok := rprt.ResolvedTaskResources.Inputs[dep.Name]
+			if !ok {
+				return fmt.Errorf("PipelineTask %s is trying to declare dependency for Resource %s which is not an input of PipelineTask %q", rprt.PipelineTask.Name, dep.Name, rprt.PipelineTask.Name)
+			}
+
+			for _, pb := range dep.ProvidedBy {
+				if pb == rprt.PipelineTask.Name {
+					return fmt.Errorf("PipelineTask %s is trying to depend on a PipelineResource from itself", pb)
+				}
+				depTask := findReferencedTask(pb, state)
+				if depTask == nil {
+					return fmt.Errorf("pipelineTask %s is trying to depend on previous Task %q but it does not exist", rprt.PipelineTask.Name, pb)
+				}
+
+				sameBindingExists := false
+				for _, output := range depTask.ResolvedTaskResources.Outputs {
```

Comment:
not 1...
not 2...
not THREE

BUT FOUR

layers of for loop nesting

:sunglasses: 

it is a beautiful thing

Reply:
:sweat_smile: 
In terms on indentantion, we could extract some loop body into functions :angel:

Reply:
hm we could! but im not sure it would be any clearer - im even thinking that making `findReferencedTask` a separate function was maybe overkill 🤔 i'm happy to do it tho if you have any ideas - the best I could think of is moving the body of the `for _, pb := range dep.ProvidedBy` loop into a function but I'm not really sure how much that improves things (and we'd have to add another layer of error handling)